### ISSUE TITLE
Active CQL requests are not waited during `nodetool drain`

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#4808]
*Installation details*
HEAD: 77686ab8890c652bb32280f537e221305fb7fb8c
Cluster size: 2

*Description*
1) Create a 2 nodes ccm cluster: 
```
ccm create scylla-2 --scylla --vnodes -n 2 --install-dir=<your scylla repo>
```
2) Populate some data:
```
cassandra-stress write n=100000 no-warmup -rate threads=300 -schema keyspace=ks12 "replication(factor=2)"
```
3) Start reading it in a loop
```
cassandra-stress read duration=2h  no-warmup -rate threads=300 -schema keyspace=ks12 "replication(factor=2)" -errors retries=0 -pop seq=1..100000
```
4) Drain one node:
```
nodetool -p 7200 drain
```
-------------------------------------------------------------------------------------

See some of read requests failing unexpectedly:
```
total,        451673,   35338,   35338,   35338,     8.5,     8.5,    20.3,    23.6,    31.0,    41.2,   14.0,  0.04803,      0,      0,       0,       0,       0,       0
total,        489537,   37864,   37864,   37864,     7.9,     7.4,    17.2,    25.0,    37.6,    40.6,   15.0,  0.04565,      0,      0,       0,       0,       0,       0
total,        521126,   31589,   31589,   31589,     9.5,     8.5,    22.4,    27.2,    30.5,    32.8,   16.0,  0.04298,      0,      0,       0,       0,       0,       0
com.datastax.driver.core.exceptions.TransportException: [127.0.0.2/127.0.0.2:9042] Connection has been closed
com.datastax.driver.core.exceptions.TransportException: [127.0.0.2/127.0.0.2:9042] Connection has been closed
com.datastax.driver.core.exceptions.TransportException: [127.0.0.2/127.0.0.2:9042] Connection has been closed
com.datastax.driver.core.exceptions.TransportException: [127.0.0.2/127.0.0.2:9042] Connection has been closed
java.io.IOException: Operation x1 on key(s) [3550374e333434313030]: Error executing: (TransportException): [127.0.0.2/127.0.0.2:9042] Connection has been closed

        at org.apache.cassandra.stress.Operation.error(Operation.java:136)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:114)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:468)
java.io.IOException: Operation x1 on key(s) [30353538354d394b4f30]: Error executing: (TransportException): [127.0.0.2/127.0.0.2:9042] Connection has been closed

        at org.apache.cassandra.stress.Operation.error(Operation.java:136)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:114)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:468)
java.io.IOException: Operation x1 on key(s) [354f4d4b4f3430363830]: Error executing: (TransportException): [127.0.0.2/127.0.0.2:9042] Connection has been closed

        at org.apache.cassandra.stress.Operation.error(Operation.java:136)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:114)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:468)
total,        524869,   24837,   24810,   24810,    11.6,     8.0,    33.1,    36.5,    38.1,    39.6,   16.1,  0.04048,      4,      0,       0,       0,       0,       0
com.datastax.driver.core.exceptions.TransportException: [127.0.0.2/127.0.0.2:9042] Connection has been closed
com.datastax.driver.core.exceptions.TransportException: [127.0.0.2/127.0.0.2:9042] Connection has been closed
java.io.IOException: Operation x1 on key(s) [34304c4f504b30315030]: Error executing: (TransportException): [127.0.0.2/127.0.0.2:9042] Connection has been closed

        at org.apache.cassandra.stress.Operation.error(Operation.java:136)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:114)
com.datastax.driver.core.exceptions.TransportException: [127.0.0.2/127.0.0.2:9042] Connection has been closed
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:468)
java.io.IOException: Operation x1 on key(s) [344b384e3431324e3230]: Error executing: (TransportException): [127.0.0.2/127.0.0.2:9042] Connection has been closed

        at org.apache.cassandra.stress.Operation.error(Operation.java:136)
        at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:114)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
        at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
        at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:468)
java.io.IOException: Operation x1 on key(s) [4f4f4d4c35354d393231]: Error executing: (TransportException): [127.0.0.2/127.0.0.2:9042] Connection has been closed
......
```
```

**Comment by glommer** [comment]
Thanks Vlad.

Some comments on my own: We have many suspicious places that I believe are related to this.
The drain code essentially calls `stop()` in the cql server, where we have this:
```
future<> cql_server::stop() {
    _stopping = true;
    size_t nr = 0;
    size_t nr_total = _listeners.size();
    clogger.debug("cql_server: abort accept nr_total={}", nr_total);
    for (auto&& l : _listeners) {
        l.abort_accept();
        clogger.debug("cql_server: abort accept {} out of {} done", ++nr, nr_total);
    }
    auto nr_conn = make_lw_shared<size_t>(0);
    auto nr_conn_total = _connections_list.size();
    clogger.debug("cql_server: shutdown connection nr_total={}", nr_conn_total);
    return parallel_for_each(_connections_list.begin(), _connections_list.end(), [nr_conn, nr_conn_total] (auto&& c) {
        return c.shutdown().then([nr_conn, nr_conn_total] {
            clogger.debug("cql_server: shutdown connection {} out of {} done", ++(*nr_conn), nr_conn_total);
        });
    }).then([this] {
        return std::move(_stopped);
    });
}
```

We don't call into the storage proxy at all, and what we have instead is a gate waiting for connections to finish.

However, the gate itself is not waited for: we call `stop()` on the gate at the end of connection processing, but we don't wait at all for the future that this returns. This is how we invoke `process()`:

```
            conn->process().then_wrapped([this, conn] (future<> f) { ...  }); <=== who waits for me??
```

Not that it would matter: we enter the gate inside `process_request()`, but leave it through a `defer` lambda that is moved as the code progresses. The last place for the move is:

```
                }().then_wrapped([this, buf = std::move(buf), mem_permit = std::move(mem_permit), leave = std::move(leave)] (future<processing_result> response_f) {
                  try {
                    auto response = response_f.get0();
                    update_client_state(response); 
                    write_response(std::move(response.cql_response), _compression);
                  } catch (...) {
                    clogger.error("request processing failed: {}", std::current_exception());
                  }
                });
```

`write_response`, however, also can block, but we don't wait for it at all:

```
    _ready_to_respond = _ready_to_respond.then([this, compression, response = std::move(response)] () mutable {
        auto message = response->make_message(_version, compression);
        message.on_delete([response = std::move(response)] { });
        return _write_buf.write(std::move(message)).then([this] {
            return _write_buf.flush();
        });
    });
```

So if we are not ready to respond right now, we may have already destroyed the connection by the time we are.

I suspect that the `rpc` code will have similar issues -  during drain we should gracefully wait for what other nodes are doing too - but I haven't much checked that.

**Comment by glommer** [comment]
@slivne please assign someone to end the life of this nefarious bug.

**Comment by roydahan** [comment]
I have a question, if c-s is set to the default retries of 10, does the retried read can/will reach to a different node during the retry while the original node is being drained?

**Comment by psarna** [comment]
OK, I did some digging and I have two partial conclusions:
1. Deferring the gate closing even further, as per Glauber's suggestion, is definitely needed, but it doesn't directly solve the issue - after fixing it, broken connections can still be observed in c-s.
2. The direct cause could be that `connection::shutdown()` unconditionally shuts down the socket descriptors, thus breaking client's connections. My initial experiment was to wait for `_pending_requests_gate` to be closed and only then shut the sockets down, but it's still not enough - even if the sockets are shut down after `_pending_requests_gate` is guaranteed to be empty, I can still observe the same kind of exceptions on c-s side. Perhaps the gate is still closed prematurely and there are more continuations where it should be entered and left correctly.

I'll continue digging tomorrow.

**Comment by avikivity** [comment]
Do we tell the client not to send more requests to this connection? Is it even possible?

Without it, there is a race between us closing the connection and the client initiating a new request.

**Comment by psarna** [comment]
@avikivity perhaps that's why I still encountered the problem (and, occasionally, a `gate closed` exception) after I experimentally added more gate-waiting. If that's the case and if there's no way to let the client know, I think it would still make sense to push my patch above that prolongs the gate for correctness, but my second idea, adding a gate-wait before shutting down the socket will just change the symptoms without fixing the cause - do you still consider it worth pushing? I guess the only effect client-side would be that they'd encounter `gate closed` errors instead of broken connections.

**Comment by glommer** [comment]
I thought we could make sockets unidirectional by closing just one side of it? or am I confused?

**Comment by psarna** [comment]
edit: after the connection is closed the client-side error would naturally be the same, but before that they're more likely to encounter some `gate closed` errors for newly arriving requests.

**Comment by psarna** [comment]
@glommer I don't know if we can, but even so, wouldn't the race stay the same from the user's perspective? It would be client sending something vs Scylla making the socket unidirectional.

**Comment by avikivity** [comment]
Yes, which end would we close?

What's needed is something like
 S: this connection is going down
 C: closes write end
 S: oh, my read end was closed. Will let in-flight requests complete and close the write end.
 C: oh, the read end was closed. How extraordinary! Will drop the connection object.

@psarna does Cassandra experience the same problems?

**Comment by psarna** [comment]
I'll verify in the office tomorrow morning

**Comment by roydahan** [comment]
I think @slivne tested it (the original behavior) with Cassandra and said that it behaves well.

It’s quite simple why we didn’t detects it in QA, we don’t test with retries=0.
CL=ONE and retires=0 sounds to me like one doesn’t really care if read/write really reach the db or not, but that’s me.

**Comment by glommer** [comment]
On Wed, Aug 7, 2019 at 1:31 PM Roy Dahan <notifications@github.com> wrote:

> I think @slivne <https://github.com/slivne> tested it (the original
> behavior) with Cassandra and said that it behaves well.
>
> It’s quite simple why we didn’t detects it in QA, we don’t test with
> retries=0.
> CL=ONE and retires=0 sounds to me like one doesn’t really care if
> read/write really reach the db or not, but that’s me.
>

that's certainly you, because that's not what the contract for CL=ONE says.

That's specially true since we're talking about a graceful operation.

**Comment by vladzcloudius** [comment]
@psarna I think your patch above doesn't hold the gate object long enough - it should be holding up until write_buf.flush() resolves (namely you need to add a `flush().finaly([leave=std::move(leave)]{}) ` clause. 

I also commented on the patch in the github.

@avikivity @glommer @psarna I think if node declares itself as DOWN driver should detect this and stop sending it packets and eventually close connections to it. But this has to be verified. 
I don't see any other way that it may possibly work due to that race @avikivity has mentioned.

**Comment by psarna** [comment]
@vladzcloudius yeah, the patch here is a little obsolete, I already tried exactly that yesterday (holding after the flush), and the same errors could still be observed. Later today, once I'm in the office, I'll  push the proper version.

**Comment by psarna** [comment]
update: c-s reports loads of errors during drain on Cassandra as well, a different kind though:
~~~
com.datastax.driver.core.exceptions.ConnectionException: [localhost/127.0.0.1:9042] Pool is shutdown
	at com.datastax.driver.core.HostConnectionPool.closeAsync(HostConnectionPool.java:613) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.SessionManager.removePool(SessionManager.java:400) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.SessionManager.onDown(SessionManager.java:485) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.Cluster$Manager.onDown(Cluster.java:1941) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.Cluster$Manager.access$1200(Cluster.java:1385) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.Cluster$Manager$5.runMayThrow(Cluster.java:1898) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.ExceptionCatchingRunnable.run(ExceptionCatchingRunnable.java:32) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [na:1.8.0_211]
	at java.util.concurrent.FutureTask.run(Unknown Source) [na:1.8.0_211]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [na:1.8.0_211]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [na:1.8.0_211]
	at com.datastax.shaded.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at java.lang.Thread.run(Unknown Source) [na:1.8.0_211]
~~~

**Comment by slivne** [comment]
Strange: I tried to run the exact test that vlad did on cassandra 3.11.3
and it worked ok - no single error emitted for cassandra

**Comment by psarna** [comment]
I did 3.11.0 (via their original ccm), but I'll double check on 3.11.3

**Comment by psarna** [comment]
Actually, on 3.11.3 I observe identical behaviour :)
~~~
java.io.IOException: Operation x1 on key(s) [4f324e383833334f3631]: Error executing: (TransportException): [localhost/127.0.0.1:9042] Connection has been closed

com.datastax.driver.core.exceptions.TransportException: [localhost/127.0.0.1:9042] Connection has been closed	at org.apache.cassandra.stress.Operation.error(Operation.java:136)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:114)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:468)
java.io.IOException: Operation x1 on key(s) [324d5031314d364c4e30]: Error executing: (TransportException): [localhost/127.0.0.1:9042] Connection has been closed

~~~

**Comment by slivne** [comment]
I'll try to double check on my end - I used scylla ccm - yet was using the
cassandra 3.11.3 for sure !!!! - which means its cassandra's code and tools

**Comment by psarna** [comment]
Ok. For reference, I used this ccm: https://github.com/riptano/ccm
and this command to create the cluster:
~~~
./ccm create cassandra-1 --vnodes -n 2 -v 3.11.3
~~~

**Comment by slivne** [comment]
I used the following


maybe the issue is the cassandra-stress are you using scylla's c-s or
cassandra ?

for reference I ran (now again)

ccm create cas-2-2 --vnodes -n 2 --version=3.11.3

ccm stress write n=100000 no-warmup -rate threads=300 -schema keyspace=ks12
"replication(factor=2)"

ccm stress read duration=2h  no-warmup -rate threads=300 -schema
keyspace=ks12 "replication(factor=2)" -errors retries=0 -pop seq=1..100000

windows 2

ccm node1 nodetool drain

 ccm node2 nodetool status

Datacenter: datacenter1
=======================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address    Load       Tokens       Owns (effective)  Host ID
                    Rack
DN  127.0.0.1  35.4 MiB   256          100.0%
 e20f804a-3366-49c3-8fed-a062e03543f4  rack1
UN  127.0.0.2  35.42 MiB  256          100.0%
 9aa57284-4dc8-4e65-a6bc-4b3f2e341a56  rack1

**Comment by psarna** [comment]
It appears so! When using `cassandra-stress` from 3.11.3, Scylla works correctly after https://github.com/scylladb/scylla/pull/4818 . I didn't verify yet if it's also the case before it.

**Comment by psarna** [comment]
Cassandra 3.11.3 works fine during drains with `cassandra-stress` from 3.11.3 as well.

**Comment by slivne** [comment]
so basiclly we need to make drain better in scylla, 

but ALSO understand why one application is behaving better than the other with drain (and that may be important for application developers ...).

Which version of scylla cassandra-stress (scylla-tools-java) are you running

**Comment by psarna** [comment]
I used https://github.com/scylladb/scylla-tools-java/commit/afb185721015105718f86e6f0b12cfc334c97645 , I'll update to master now.

**Comment by avikivity** [comment]
Perhaps we should do something like
1. send a DOWN event
2. wait 300ms for the client to digest this
3. close the connection, if the client hasn't done so on its own.

We should also check the drivers to see whether they close the connection on their own. If they all do, then we can increase the 300ms timeout.

**Comment by psarna** [comment]
scylla-tools-java master is also affected

**Comment by slivne** [comment]
I think we did the a similar analysis on how cassandra shutdown's more cleanly then us Internal email "Scylla vs Cassandra shutdown code"

My bet is that cassandra sends an RST not a FIN on the sockets when closing and during drain - we have discussed this in the past (https://github.com/apache/cassandra/blob/8a04204c10eafc2c1051a336d82211899889e276/src/java/org/apache/cassandra/transport/Server.java#L131 - using SO_LINGUR) - so they are reducing the window in which the client will send requests when drain is requested.
(that does not mean they do not process the requests they have received)

I also think we figured out the difference in drivers - we are using a newer version because we needed additional features here part of scylladb/scylla#3843 - https://github.com/scylladb/scylla/issues/3843#issuecomment-433050953

To summarize:
- we can make process the requests already received
- scylla and cassandra have issues when c-s is using the newer driver (to bad ...) - seems awful that only old driver work, lets try upstream I guess.
- cassandra probably forces the client to close the connection and not send additional requests during drain more forcefully - by not sending a TCP FIN but sending a TCP RST and scylla probably doesn't do that.

**Comment by avikivity** [comment]
How will RST help? We will not be able to respond to any ongoing requests.

**Comment by avikivity** [comment]
@psarna this is closed by 01410c9770c51c37bccde070c29a533227bf7a80? or is more work needed?

**Comment by psarna** [comment]
I finished my analysis with the statement that new `cassandra-stress` from stock C* 3.11.3 doesn't suffer from this, both with Scylla and C*, while our c-s has the same symptoms on both Scylla and C* 3.11.3, which suggests that's a c-s issue. I haven't tried reproducing after 01410c9, I can do that tomorrow.

**Comment by psarna** [comment]
01410c9 seems neutral to the issue, the symptoms are exactly the same.

**Comment by slivne** [comment]
@psarna we need to go back to this and understand whats going and why we have different beavior with the scylla c-s look here for info on the difference between versions https://github.com/scylladb/scylla/issues/4808#issuecomment-519588734

**Comment by psarna** [comment]
It's been a while, so I checked again, and first of all I can confirm that the issue is with c-s only, because of these:

* scylla cluster + scylla-tools-java's cassandra stress -> doesn't work properly
* scylla cluster + stock 3.11.3 cassandra stress -> works fine
* cassandra 3.11.3 cluster + scylla-tools-java's cassandra stress -> doesn't work properly
* cassandra 3.11.3 cluster + stock 3.11.3 cassandra stress -> works fine

I also tried injecting the newer driver to 3.11.3 cassandra stress (as per https://github.com/scylladb/scylla/issues/3843#issuecomment-433050953) in order to see if it's the newer driver's fault, but I ended up with
```
	at java.lang.Thread.run(Unknown Source) [na:1.8.0_211]
ERROR 13:57:30,454 Unexpected error while querying localhost/127.0.0.1
com.datastax.driver.core.exceptions.ConnectionException: [localhost/127.0.0.1:9042] Pool is shutdown
	at com.datastax.driver.core.HostConnectionPool.closeAsync(HostConnectionPool.java:613) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.SessionManager.removePool(SessionManager.java:400) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.SessionManager.onDown(SessionManager.java:485) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.Cluster$Manager.onDown(Cluster.java:1941) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.Cluster$Manager.access$1200(Cluster.java:1385) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.Cluster$Manager$5.runMayThrow(Cluster.java:1898) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at com.datastax.driver.core.ExceptionCatchingRunnable.run(ExceptionCatchingRunnable.java:32) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [na:1.8.0_211]
	at java.util.concurrent.FutureTask.run(Unknown Source) [na:1.8.0_211]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [na:1.8.0_211]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [na:1.8.0_211]
	at com.datastax.shaded.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [cassandra-driver-core-3.5.1-shaded.jar:na]
	at java.lang.Thread.run(Unknown Source) [na:1.8.0_211]
```
which may have more to do with the fact that I just copy-pasted a .jar file. Nonetheless, the error above occurred only on `nodetool drain`, the reading process itself worked without issues.

**Comment by slivne** [comment]
To elaborate - the concern is that this is not confined to
cassandra-stress, scylla users are using the newer java driver (since the
shard aware version is based on that).

So all users using the shard-aware driver can get this errors when they
drain.

We want to validate if Scylla's drain can be updated to make this more
"clean".

**Comment by psarna** [comment]
For testing purposes, I added the following to `storage_service::drain()`:
1. drain is initiated
2. DOWN event for this node is pushed to all listeners (via a STATUS_CHANGE event, using the event_notifier from cql_server::connection)
3. wait a while for the client to acknowledge the event
4. continue with the drain (shut down the sockets, etc.)

I verified via debug printing that the status change is indeed sent to somebody - and since there's only one active CQL connection in my test case, I assume it was sent to cassandra-stress [*]. Still, even after waiting a couple of seconds, once the code starts continuing the drain and shutting down the sockets, cassandra-stress reports the same "Connection broken" errors. The highest time I checked was 10 seconds, which should be 50x enough for a local cassandra-stress instance to acknowledge that the node is going down. Perhaps this driver ignores this event?

[*] - verified via Wireshark while writing this post - an `S->C` event was indeed pushed with a status change to `DOWN`

**Comment by psarna** [comment]
From a comment in https://github.com/datastax/java-driver/blob/9f0d89799a8a1e4cd1022dd7c43333924c36a648/core/src/main/java/com/datastax/oss/driver/internal/core/metadata/TopologyEvent.java#L73 :

if the node is currently ignored by the driver's load balancing policy, this is reflected
in the driver metadata's corresponding {@link Node}, for information purposes only.
**otherwise, if the driver still has at least one active connection to that node, this is
ignored**. In other words, a functioning connection is considered a more reliable
indication than a topology event.
<p>If you want to bypass that behavior and force the node down, use {@link
#forceDown(InetSocketAddress)}.

So, looks like the driver thinks it knows better :) Without patching the driver there's likely no way of fooling it into thinking the connection is down, if the connection itself works.

**Comment by psarna** [comment]
update: when I hardcode the driver to be less self-confident and accept my authority when I send it a "node is down" status, then sending the `DOWN` event works kind of fine - it's still possible to see a single "Connection closed" sometimes (although the connection remains open Scylla-side, so the driver must have closed it by itself), but after a single error like that (or, sometimes, no error at all, so it's a race in the driver) the driver acknowledges that the node is really down and stops issuing new requests.

Still, it's not really a viable solution, since it requires a driver patch. Also, it would probably make sense to leave the current semantics of `STATUS_CHANGE:DOWN` as is, in case somebody relies on it, so the solution would involve adding another `STATUS_CHANGE` verb. Not sure if it's worth the trouble.

By the way - I also tested with a much fresher original driver (4.3.0), because they reworked most of the code, but this issue stayed the same.

**Comment by vladzcloudius** [comment]
@psarna I agree. This doesn't look worth the trouble of changing anything on our side.

**Comment by slivne** [comment]
@psarna I am going to insist we do something about this - having drain fail requests defeats the purpose of using drain in the first place.

Drain is supposed to be the safe way to stop traffic 

cassandra 3.11 uses old drivers 3.0.X https://github.com/apache/cassandra/blob/cassandra-3.11/build.xml
cassandra 4 uses a newer driver 3.6.0 https://github.com/apache/cassandra/blob/trunk/build.xml#L560

can you take this for a spin on scylla and cassandra, AFAIK 3.6.0 is new enough to show the issues for scylla and cassandra.

If Cassandra 4 is not showing them I want to understand why - what have they done.

**Comment by psarna** [comment]
Sure, I'll compare Cassandra 4, Cassandra 3.11 and Scylla with 3.6.0 and post results.

**Comment by psarna** [comment]
Alright, I checked out Cassandra 4 (trunk, which is at version 4-alpha3). They indeed use 3.6.0 driver in their cassandra-stress, here's a confirmation:

```cql

cqlsh> select * from system_views.clients ;

 address   | port  | connection_stage | driver_name          | driver_version | hostname  | protocol_version | request_count | ssl_cipher_suite | ssl_enabled | ssl_protocol | username
-----------+-------+------------------+----------------------+----------------+-----------+------------------+---------------+------------------+-------------+--------------+-----------
 127.0.0.1 | 38240 |            ready |                 null |           null | localhost |                4 |            30 |             null |       False |         null | anonymous
 127.0.0.1 | 38242 |            ready |                 null |           null | localhost |                4 |            34 |             null |       False |         null | anonymous
 127.0.0.1 | 38282 |            ready | DataStax Java Driver |          3.6.0 | localhost |                4 |            15 |             null |       False |         null | anonymous
 127.0.0.1 | 38284 |            ready | DataStax Java Driver |          3.6.0 | localhost |                4 |             3 |             null |       False |         null | anonymous
 127.0.0.1 | 38286 |            ready | DataStax Java Driver |          3.6.0 | localhost |                4 |             2 |             null |       False |         null | anonymous
 127.0.0.1 | 38288 |            ready | DataStax Java Driver |          3.6.0 | localhost |                4 |             2 |             null |       False |         null | anonymous
 127.0.0.1 | 38290 |            ready | DataStax Java Driver |          3.6.0 | localhost |                4 |             2 |             null |       False |         null | anonymous
 127.0.0.1 | 38292 |            ready | DataStax Java Driver |          3.6.0 | localhost |                4 |             
```

Both Cassandra4 (spun up with original ccm) and Scylla clusters (spun up with scylla-ccm) and both end up with exactly the same error.

Cassandra4:
```
	at org.apache.cassandra.stress.Operation.error(Operation.java:127)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:105)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:91)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:99)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:248)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:466)
java.io.IOException: Operation x1 on key(s) [504b354b394e4f4f3131]: Error executing: (TransportException): [localhost/127.0.0.1:9042] Connection has been closed

	at org.apache.cassandra.stress.Operation.error(Operation.java:127)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:105)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:91)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:99)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:248)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:466)
java.io.IOException: Operation x1 on key(s) [4f4c36504d36374c3230]: Error executing: (TransportException): [localhost/127.0.0.1:9042] Connection has been closed
```

Scylla:
```
	at org.apache.cassandra.stress.Operation.error(Operation.java:127)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:105)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:91)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:99)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:248)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:466)
java.io.IOException: Operation x1 on key(s) [4d323536503134333631]: Error executing: (TransportException): [localhost/127.0.0.1:9042] Connection has been closed

	at org.apache.cassandra.stress.Operation.error(Operation.java:127)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:105)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:91)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:99)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:248)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:466)
java.io.IOException: Operation x1 on key(s) [37333933323935323531]: Error executing: (TransportException): [localhost/127.0.0.1:9042] Connection has been closed
```

**Comment by slivne** [comment]
Thanks 

So I think we can at least remove the Field-Tier1 item from it.

Still we should include instructions to applications how to handle this and find a way to make it better. There are operations that are not idempotent and as such it would be "better" to find a way to allow doing drains in a manner where such errors would not happen.
